### PR TITLE
test: Support pytest-style fixtures

### PR DESCRIPTION
**Pull request by dcramer**
- Move Fixtures into Factories (create_X helpers)
- Add pytest class-based fixture for unittest compatibility
- Move several database TestCase tests to standard unit tests (didnt require a database)

### CODE REVIEW

**Comment by dcramer** [comment]
This adopts the pattern I implemented in Zeus (which was modeled off of the pattern I created in Sentry), which was generally fairly easy to use. It will let you do function-based tests, which unlocks the ability to use parameterization and other advanced pytest features.

Probably failing CI at the moment, but wanted to get this up for feedback. I'll sort out the test failures.

File: src/sentry/testutils/cases.py [hunk 1]
```diff
@@ -76,6 +73,27 @@
 DEFAULT_USER_AGENT = 'Mozilla/5.0 (Windows NT 6.1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/41.0.2228.0 Safari/537.36'
 
 
+@pytest.fixture(scope="function")
```
Comment (dcramer) [inline src/sentry/testutils/cases.py]: i should probably move these to the unittest plugin now

Comment (dcramer) [reply src/sentry/testutils/cases.py]: jk i did and this is duplicated

File: src/sentry/utils/pytest/fixtures.py [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+
+    return Factories
+
+
+@pytest.fixture(scope='function')
+def session():
+    return factories.create_session()
+
+
+@pytest.fixture(scope='function')
```
Comment (dcramer) [inline src/sentry/utils/pytest/fixtures.py]: these are all copypasta but it also decouples old style from new style meaning we could create or augment better defaults for newer tests

**Comment by dcramer** [comment]
```python
def look_ma_test(db, default_user):
    assert default_user.email == 'default@example.com'
```

**Comment by dcramer** [comment]
One thing to give thought here - it'd be nice if we could specify "needs db" and "transactional_db" is an upgrade:

https://pytest-django.readthedocs.io/en/latest/helpers.html#transactional-db

Its unclear if thats actually how things work, but then it means we could imply default_X fixtures all require the db, which gets us closer to being able to run unittests without setting up the db (unless required).

**Comment by dcramer** [comment]
jk:

>     If both this and ``transactional_db`` are requested then the
>     database setup will behave as only ``transactional_db`` was
 >    requested.

File: tests/sentry/test_canonical.py [hunk 3]
```diff
@@ -2,13 +2,14 @@
...
+import unittest
+
 from collections import OrderedDict
 
-from sentry.testutils import TestCase
 from sentry.utils.canonical import CanonicalKeyView, CanonicalKeyDict
 
 
-class CanonicalKeyViewTests(TestCase):
+class CanonicalKeyViewTests(unittest.TestCase):
```
Comment (untitaker) [inline tests/sentry/test_canonical.py]: If you rename `FooTests` to `TestFoo`, you don't need to inherit from `TestCase` to get the tests discovered. Then using fixtures on each test method should just work

**Review by untitaker — APPROVED** [review]

File: tests/sentry/test_canonical.py (hunk 3 above)
Comment (dcramer) [reply tests/sentry/test_canonical.py]: I actually only changed these so we could avoid the database activity since they're purely python code tests

**Pull request merged**